soc/st/stm32/common/stm32_wkup_pins.c

```diff
@@ -0,0 +1,345 @@
+		.gpios_cfgs =                                                                      \
+		{                                                                                  \
+		FOR_EACH_FIXED_ARG(WKUP_PIN_GPIOS_CFG_DT_BY_IDX, (,), node_id,                     \
+			LISTIFY(PWR_STM32_WKUP_PIN_SRCS_NB, NB_DEF, (,)))                          \
+		}                                                                                  \
+	}
+
+/* wkup_pin idx starts from 1 */
+#define WKUP_PIN_CFG_DT_BY_IDX(idx) WKUP_PIN_CFG_DT(WKUP_PIN_NODE_ID_BY_IDX(idx))
+
+#define PWR_STM32_WKUP_PIN_DEF(i, _) LL_PWR_WAKEUP_PIN##i
+
+/**
+ * @brief Structure for passing the configuration of a wake-up pin.
+ */
+struct wkup_pin_cfg_t {
+	uint32_t wkup_pin_id;
+#if PWR_STM32_WKUP_PINS_POLARITY
+	bool polarity; /* True if detection on the low level : falling edge */
+#endif                 /* PWR_STM32_WKUP_PINS_POLARITY */
+#if PWR_STM32_WKUP_PINS_PUPD_CFG
+	uint32_t pupd_cfg;      /* wake-up pin GPIO port pull-up/pull-down config */
+	uint32_t gpio_port;    /* wake-up pin GPIO port */
+	gpio_pin_t gpio_pin;   /* wake-up pin GPIO pin number */
+#endif /* PWR_STM32_WKUP_PINS_PUPD_CFG */
+	uint32_t src_selection; /* The source signal to use with this wkup pin */
+};
+
+/*
+ * LookUp Table to store the LL_PWR_WAKEUP_PINx for each pin.
+ */
+static const uint32_t table_wakeup_pins[PWR_STM32_MAX_NB_WKUP_PINS + 1] = {
+	LISTIFY(PWR_STM32_MAX_NB_WKUP_PINS, PWR_STM32_WKUP_PIN_DEF, (,))};
+
+#define WKUP_PIN_CFG_DT_COMMA(wkup_pin_id) WKUP_PIN_CFG_DT(wkup_pin_id),
+
+static struct wkup_pin_dt_cfg_t wkup_pins_cfgs[] = {
+	DT_FOREACH_CHILD(STM32_PWR_NODE, WKUP_PIN_CFG_DT_COMMA)};
+
+#if PWR_STM32_WKUP_PINS_PUPD_CFG
+
+#define STM32_GPIOS_NB               16
+
+#if defined(CONFIG_SOC_SERIES_STM32U5X) || defined(CONFIG_SOC_SERIES_STM32WBAX)
+#define PWR_STM32_GPIO_BIT_DEF(i, _) LL_PWR_GPIO_PIN_##i
+#else
+#define PWR_STM32_GPIO_BIT_DEF(i, _) LL_PWR_GPIO_BIT_##i
+#endif /* CONFIG_SOC_SERIES_STM32U5X or CONFIG_SOC_SERIES_STM32WBAX */
+
+/** @endcond */
+
+/*
+ * LookUp Table to store the LL_PWR_GPIO_BIT_x for each pin.
+ * In case of U5 & WBA series, it's LL_PWR_GPIO_PIN_x
+ */
+static const uint32_t table_pwr_gpio_bits[STM32_GPIOS_NB] = {
+	LISTIFY(STM32_GPIOS_NB, PWR_STM32_GPIO_BIT_DEF, (,))};
+
+/**
+ * @brief Array containing pointers to each GPIO port.
+ *
+ * Entries will be NULL if the GPIO port is not enabled.
+ */
+static const struct device *const gpio_ports[] = {
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpioa)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpiob)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpioc)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpiod)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpioe)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpiof)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpiog)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpioh)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpioi)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpioj)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpiok)),
```

Comment:
You could use macros for checking if the give nodelabel is enabled and save a few array entries on SoCs with fewer GPIO ports.

Reply:
In addition to getting the device ptr of each GPIO port, this table is also used to get the index of a given GPIO port, so keeping the NULL entries is desired.